### ISSUE TITLE
[Bug]: Too many disk writes on update

### ISSUE THREAD

**Issue by mcrha** [flatpak/flatpak#5982]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.15.10

### What Linux distribution are you using?

Fedora Silverblue

### Linux distribution version

41

### What architecture are you using?

x86_64

### How to reproduce

Run apps update through gnome-software and check how many bytes had been written to the disk by the app.

### Expected Behavior

too many bytes written to the disk, can be terabytes

### Actual Behavior

hopefully less than terabytes on a single update of many Flatpak apps

### Additional Information

_To expand on the above short text..._

This had been reported against the gnome-software:  
https://gitlab.gnome.org/GNOME/gnome-software/-/issues/2705

The user noticed that the gnome-software process had 2.3TB of the disk writes, which is quite a lot.

After some investigation I believe it's due to the Flatpak updates, which are executed within the process (through the flatpak plugin in the gnome-software). The reporter has installed many flatpak apps, thus if many had an update, and/or the runtimes, then it makes sense to have many writes, but even with that the 2.3TB looks unbelievable (from my point of view). I can imagine there will be more and more flatpaks, some distros won't use anything else, then it'll write how many, 10TB? As had been said in the gnome-software bug, it might not be a big deal for the HDD, but it could kill the SSD more or less soon.

This is a plea to check whether all the writes are necessary and whether there's a way to save some of the writes, to keep users' disk live longer. It can be a problem in a library(ies) the flatpak uses, but it's up to you to decide, because I do not know the flatpak internals.

**Comment by francoism90** [comment]
@mcrha Thanks for the help for reporting this issue. :)

It indeed seems to be caused by Flatpak and/or combination with ostree.

This seems to be working as it should the last few days (almost no writes). But if any command or debug info would be useful, please let me know!

**Comment by varunbpatil** [comment]
Came here from the gnome-software gitlab issue. I have the same problem. I just had gnome-software write 3.5TB in a couple of hours of uptime. I'm not running Silverblue or any other atomic distro. I'm running Arch with Gnome 47. I had the following flatpak update:
```sh
❯ flatpak history
...
Dec 13 11:39:47 deploy update  org.freedesktop.Platform.GL.nvidia-565-77 1.4    system       flathub
```

This seems like a very serious issue for ssd longevity, unless the disk write stats that tools like bottom are reporting are wrong.

**Comment by varunbpatil** [comment]
I was able to recreate this issue by uninstalling and re-installing zoom flatpak which installs `org.freedesktop.Platform.GL.nvidia-565-77`. 

I was watching the total disk writes and it instantly jumped to 3.5 TB when installing `org.freedesktop.Platform.GL.nvidia-565-77`. 

I doubt my drive is capable of writing that many terabytes in a few seconds. So, definitely must be something else.
Could btrfs be the issue?

**Comment by francoism90** [comment]
@varunbpatil I don't know why, but it seems to write a lot. Like 50GB in a few hours for me. Which is really weird when not doing much:

![image](https://github.com/user-attachments/assets/c1242421-06ca-4229-a76c-51fe9aacda19)

I understand I/O is needed, but it feels like a leak somewhere.

**Comment by varunbpatil** [comment]
@hfiguiere My comment above regarding recreating with `org.freedesktop.Platform.GL.nvidia-565-77` is without gnome-software. It's directly with flatpak cli.

**Comment by francoism90** [comment]
@hfiguiere Please read https://gitlab.gnome.org/GNOME/gnome-software/-/issues/2705#note_2260748

I have reported this bug to gnome-software, which upstream think is caused by flatpak itself.

I don't know what's the cause. :)

**Comment by smcv** [comment]
> I was watching the total disk writes and it instantly jumped to 3.5 TB when installing `org.freedesktop.Platform.GL.nvidia-565-77`.
> 
> I doubt my drive is capable of writing that many terabytes in a few seconds.

If a diagnostic tool seems to be reporting something that isn't physically possible, then either the diagnostic tool is wrong, or someone or something is misinterpreting its output. Flatpak cannot magically make your SSD faster!

For example, if `flatpak-system-helper` is doing a lot of writes to a tmpfs or to a virtual file system like FUSE, that would be one way to see a large number of filesystem writes that do not actually do anywhere near that much disk I/O. I know that the system helper *does* use a FUSE filesystem (`revokefs-fuse`) as part of its security boundary between users, so it's possible that that mechanism confuses the disk write statistics.

If nobody can identify a situation where Flatpak (or libostree or revokefs-fuse) is genuinely doing a lot of unnecessary disk writes, then this issue report is not going to be actionable.

> a very serious issue for ssd longevity

If you are concerned about SSD longevity, the disk write statistics to be watching are the ones reported by disk-level tools like `smartctl -a /dev/nvme0n1`.

**Comment by varunbpatil** [comment]
@smcv I agree. If it is not within the drive's capabilities, it must be something else. I did check `smartctl -a /dev/nvme0n1` before and after and _did not_ see a difference of 3.5 TB (assuming "Data Units Written" is what I should be looking at). So, yes, it might just be bad reporting by these tools. But, should be easy enough to confirm by someone more knowledgeable.

**Comment by francoism90** [comment]
@smcv This issue was created to find out the cause, and even to provide more debugging. :)

If you or anyone have a good recommendation of things to diagnose, please let me know. I didn't check smartctl for example.

**Comment by mcatanzaro** [comment]
This issue is tagged "needs more info" but there's not actually a request for more info here. We need to know which additional info is required.
